`pytest`

JacksonBurns opened the pull request:
See #2380

JacksonBurns commented:
This PR will also resolve #975 - already in the spirit of cleaning up dead tests, so it is coming along naturally.

rwest commented on test/database/databaseTest.py:
The reason we did these as generators is so that you get a test for each family, and it doesn't just check all the families in one test and then fail at the first error. Can we preserve some behavior like that with pytest?

JacksonBurns replied:
For myself in the future: https://github.com/okken/pytest-check

JacksonBurns commented:
## Summary of Changes for Review
Below is a list of changes in this PR, separated into 'simple' and 'not-so-simple' changes (and some chores).

### Simple Changes
 - All unit, functional, and database tests have been moved to `test/type` directory
 - The `testing` folder was removed
 - Applied formatting updates across many files
 - Switched test runner from `nosetest` to `pytest` (has feature parity, plus some nice extras)
 - Importing `julia` __works__ in the testing framework now
   - Without adding unit tests explicitly for the interconnect, coverage has gone up since `reactors.py` 'works' now

### Not-so-Simple Changes
 - `pytest` does things a bit differently than `nosetest`, I think for the better
   - I call `pytest` only once which cuts down the total time to run + a nice speedup from `pytest` being maintained still, but (see below) this means one of the tests no longer works (for this or potentially other reasons)
   - `pytest` handles execution order, namespaces, and `setup`/`teardown` methods differently than `nosetest`, and I have done my best to either reproduce _exactly_ what we used to have or 'fix' them elsewhere
 - I have removed `HinderedRotor2D` and its associated tests and examples 
   - `HinderedRotor2D` which uses `Q2DTor` and has not been tested in 2+ years (cloning and preparing `Q2DTor` never made it into GitHub actions CI). @mjohnson541's Python3 is no longer needed since the upstream has a [official Python 3 port](https://github.com/cathedralpkg/Q2DTor). 
   - Rather than carry forward a non-functional test, I opted to remove support now so that we could _potentially_ add it back with this new version.
 - `test/arkane/testExplorerJob.py:TestExplorerJob` is now skipped during testing
   - As stated in the `reason` ([see here](https://github.com/ReactionMechanismGenerator/RMG-Py/blob/24d1adc59d43e9c48dd3e2ae7bbbfcb6e5df85ae/test/arkane/explorerTest.py#L43)) this test was reliant on the global state in a way that I can't reproduce in the new framework
   - This doesn't impact the test coverage, and the example runs locally with no issues if run on its own
   - Also as mentioned in the skip, I think the best way to 'fix' this functional test is to fix the use of `global`, `exec`, etc. in Arkane instead (which I intend to do as my next big task).

### Cleanup
 - Various Windows batch scripts that we don't use or don't work removed
 - Removed methods which were deprecated before RMG-Py v3, updated docs accordingly
 - `external` folder was removed
 - Outdated `Makefile` targets removed

## Follow Up Steps
In subsequent PRs I plan to (1) fix the failing Arkane test by removing the abuse of global state in Arkane and (2) add more unit tests for the `pyjulia` interconnect. At least the latter would be in scope of this PR, but it's already a huge diff and I think keeping it separate would be better.

kspieks commented on pytest.ini:
```diff
@@ -0,0 +1,3 @@
+[pytest]
+testpaths = test
+python_files = *Test.py
```
Should a newline be added here?

kspieks commented on test/arkane/ess/arkaneGaussianTest.py:
```diff
@@ -0,0 +1,389 @@
+#                                                                             #
+# RMG - Reaction Mechanism Generator                                          #
+#                                                                             #
+# Copyright (c) 2002-2021 Prof. William H. Green (whgreen@mit.edu),           #
```
I don't remember if we have an official rule on when we update the dates at the top of each file, but just thought I’d point out that it’s 2 years old now. Probably worth a separate PR in the future

JacksonBurns replied:
I am pro removing anyway, it feels very 70's :laughing:

rwest replied:
It's part of the "new release" workflow; I think there's a script to do it.

kspieks commented on test/rmgpy/chemkinTest.py:
```diff
@@ -926,21 +926,21 @@ def setUpClass(cls):
                              Reaction index: Chemkin #2; RMG #4
                              Template reaction: H_Abstraction
-                             Estimated using template (C/H3/Cs;C_methyl) for rate rule (C/H3/Cs\H3;C_methyl)
+                             Estimated using template (C/H3/Cs;C_methyl) for rate rule (C/H3/Cs\\H3;C_methyl)
```
Just for my curiosity, why does the \ have to be escaped now?

JacksonBurns replied:
Pending deprecation in future version of Python - `\H` is being interpreted as an escape character (but `\H` is _not_), being found invalid, and then treated as literal `\H`. In the future this will error, just fixing it now

kspieks commented on test/arkane/arkaneInputTest.py:
```diff
@@ -53,11 +52,10 @@
     process_model_chemistry,
 )
 from arkane.modelchem import LevelOfTheory, CompositeLevelOfTheory
+import pytest
```
Should pytest be imported before local objects from RMG? Not sure how many files this will impact

JacksonBurns replied:
We could hit the whole repo with `isort` - there are jumbled imports all over the place. Agreed though that we could at least do it here with `pytest` to start.

kspieks reviewed (commented):
Thanks for this enormous and long-overdue improvement for RMG! My few comments center around formatting since I trust that the logic of the tests is unchanged.

rwest commented on arkane/common.py:
```diff
@@ -655,13 +933,12 @@ def get_moment_of_inertia_tensor(coords, numbers=None, symbols=None):
-        raise InputError(f'The number of atoms ({len(symbols)}) is not equal to the number of '
-                         f'atomic coordinates ({len(list(coords))})')
-    tensor = np.zeros((3, 3), np.float64)
+        raise InputError(f"The number of atoms ({len(symbols)}) is not equal to the number of " f"atomic coordinates ({len(list(coords))})")
```
```suggestion
        raise InputError(f"The number of atoms ({len(symbols)}) is not equal to the number of atomic coordinates ({len(list(coords))})")
```

rwest commented on arkane/encorr/isodesmic.py:
```diff
@@ -76,35 +76,40 @@ def __init__(self, molecule, low_level_hf298, level_of_theory, high_level_hf298=
-            raise ValueError(f'ErrorCancelingSpecies molecule attribute must be an rmgpy Molecule object. Instead a '
-                             f'{type(molecule)} object was given')
+            raise ValueError(
+                f"ErrorCancelingSpecies molecule attribute must be an rmgpy Molecule object. Instead a " f"{type(molecule)} object was given"
```
Are there any open pull requests (or other developers about to open them) touching arkane code?  A change to all the unit tests is disruptive enough - but there are many unrelated formatting changes to arkane in here too, that have nothing to do with unit tests.

JacksonBurns replied:
- #1899 is implementing new Arkane feature(s)
 - #1955 changes unit tests, but I saw your comment about that PR being confusing
 - #1967 is modifying tests and source
 - #2026 hits some arkane source
 - #2217 hits arkane source but seems abandoned

rwest reviewed (commented):
Incomplete review. I will resume later.  Is there *any* way to remove all the `'` to `"` conversions, at least in all the (not testing related) arkane code??

JacksonBurns commented on .github/workflows/CI.yml:
```diff
@@ -185,12 +185,9 @@ jobs:
-        run: make test-database
+      - name: Run Unit, Functional, and Database Tests
+        # aggregate into one command so we only have to eat the collection time once
+        run: make test-all
```
Collection time on my machine is ~5mins, so maybe 5-10 minutes on GH actions? That x3 is tiresome

Also the database test is already called testDatabase or something like that, could we name the functional test along the same lines so that they can be run together but still labeled separately?

JacksonBurns commented on arkane/common.py:
```diff
@@ -64,27 +64,28 @@
-                     'CompositeLevelOfTheory': CompositeLevelOfTheory
-                     }
+ARKANE_CLASS_DICT = {
+    "ScalarQuantity": ScalarQuantity,
```
I think for _this_ file we could just revert all the changes - I didn't change any source here. I could also presumably configure VSCode's Black to avoid this conversion. But more broadly I do think this is a good change for readability and maintenance considerations. Worth I think the extra scroll through the blame. Could be part of a larger discussion among the devs

JacksonBurns commented on .gitignore:
This was a byproduct of trying to make some of the more global state-dependent tests work. One of the tests for Arkane runs all of the examples and then deletes some of the output files, but repeat runs on my machine would fail spuriously because files were being deleted when they shouldn't be.

The state I have the `examples` directory in now includes everything that used to be there _except_ (1) an `output.py` in `arkane/reactions` that I needed locally to get passes and (2) the `CH2CHOOH` example which called HinderedRotor2D (but not Q2DTor)

JacksonBurns replied:
I could squash all the commits that touch that dir - let's see what other change end up being made and we can worry about the history at the end

rwest replied:
> we can worry about the history at the end

Code review is easier with a cleaner history.  Especially a PR that changes 90,000+ lines of code is hard to review line by line, so it's easier do some of it commit by commit. I keep having thoughts like "why is this being added (removed)?" only to find it was reverted some commits later.

JacksonBurns replied:
Ok, I will clean up some of the worst offenders now

JacksonBurns replied:
So this would undo the commits but keep the changes, and then I could re-commit? Seems like it could work. I guess I could reset back to the first problematic commit? Or we could suffer through this mega-pr and I could promise to do better next time (I do) :smiling_face_with_tear:

hwpang replied:
Yes, this keeps all the changes and allows you to recommit, and in my opinion it is much easier to recommit than dealing with rebase/fix up

rwest replied:
I have done a lot of rebasing now to tidy up the history. I suggest we force-push what's currently `rwest/pytest3` to be this `JacksonBurns/pytest` branch for the PR.

hwpang commented on test/arkane/arkaneKineticsTest.py:
```diff
@@ -62,16 +62,20 @@ def test_give_temperature_range_for_kineticsjob(self):
         self.assertEqual(50, min(kjob.Tlist.value_si))
         self.assertAlmostEqual(4000, max(kjob.Tlist.value_si))
         inverse_tlist = 1 / kjob.Tlist.value_si
-        self.assertAlmostEqual(inverse_tlist[1] - inverse_tlist[0],
```
The commit message should reflect the fact that formatting is done here as well

hwpang commented on test/rmgpy/chemkinTest.py:
```diff
@@ -993,13 +993,13 @@ def setUpClass(cls):
-            ["C/H3/Cs\H3", "C_methyl"],
-            ["C/H3/Cs\H3", "C_rad/H2/Cs"],
-            ["C/H3/Cs\H3", "H_rad"],
+            [r"C/H3/Cs\H3", "C_methyl"],
```
What is the r for? regex? Why only adding for the first one in the list?

JacksonBurns replied:
`r` stands for raw - see [this discussion](https://github.com/ReactionMechanismGenerator/RMG-Py/pull/2504#discussion_r1273868680 with @kspieks above - the `\H` is being interpreted as an escape character so the raw string tells python to not do that

hwpang commented on rmgpy/rmg/model.py:
```diff
@@ -1135,8 +1135,7 @@ def add_species_to_core(self, spec):
         if spec in self.edge.species:
-            if not self.edge.phase_system.in_nose:
-                self.edge.phase_system.pass_species(spec.label,self.core.phase_system)
+            self.edge.phase_system.pass_species(spec.label,self.core.phase_system)
```
Very happy that now we can run these tests on RMS reactors

JacksonBurns replied:
:tada: I am planning a follow up PR to add unit tests to that effect

hwpang commented on pytest.ini:
```diff
@@ -3,6 +3,7 @@
 markers =
     functional: functional tests
     database: database tests
+# need pytest-cov to do --cov=arkane,rmgpy (?)
```
What does this mean? What is it for?

JacksonBurns replied:
that was a note I left for myself and committed accidentally. a reminder that we need to have `pytest-cov` installed to run the tests. I later found that you could have a required plugins part of the config file so now it lives there

hwpang commented on examples/arkane/networks/CH2NH2/yaml_files/CH2NH.yml:
```diff
@@ -1,245 +0,0 @@
-RMG_version: 3.0.0
```
What were these example files for previously? Why remove it?

JacksonBurns replied:
Ultimately added these back, see discussion [here](https://github.com/ReactionMechanismGenerator/RMG-Py/pull/2504#discussion_r1273873059) about cleaning up the history

rwest commented on examples/arkane/reactions/NH2+N2H3=NH+N2H4/output.py:
this looks (from its file name) like an output file? should it be tracked in git?

JacksonBurns replied:
I think you are right, I will try deleting and running now

JacksonBurns replied:
This was right, file gone now

rwest commented:
Please could you update http://reactionmechanismgenerator.github.io/RMG-Py/users/rmg/installation/anacondaDeveloper.html#debugging in this PR?
(and perhaps also paste the new suggestion as a comment to help PR reviewers?)

JacksonBurns commented:
I have fixed the initial review comments in 36e0ff3 with the exception of the issues with the git history - see my comment [here](https://github.com/ReactionMechanismGenerator/RMG-Py/pull/2504#discussion_r1275081444).

rwest commented:
If anyone is trying to review this, keep an eye on my [`pytest3`](https://github.com/ReactionMechanismGenerator/RMG-Py/compare/main...rwest:RMG-Py:pytest3) branch. Beware of churn - I am force pushing quite a lot to it - but I am working to clean up the git history, which may make it easier to review. The net changes should end up similar to this pytest branch, but trying to avoid adding and removing things, or removing and adding things, or breaking and fixing things, as much as possible.

rwest commented:
I'm not sure if it's a consequence of running all the tests is one go (unit, database, and functional) or if it's pytest, or if there's some other way to avoid it, but several of my CI runs (on push) have cancelled in a way that's symptomatic of a test runner machine running out of memory.

https://github.com/rwest/RMG-Py/actions/runs/5695453110/job/15438631460#step:10:13633
make: *** [Makefile:61: test-all] Killed
Error: The operation was canceled.
soon after logging
running methyl+formaldehyde
Arkane execution initiated at Fri Jul 28 20:14:08 2023

https://github.com/rwest/RMG-Py/actions/runs/5688133570/job/15417571070#step:10:13633
Loading thermodynamics group database from /home/runner/work/RMG-Py/RMG-database/input/thermo/groups...
make: *** [Makefile:61: test-all] Killed
Error: The operation was canceled.
shortly after logging 
running methyl+formaldehyde
Arkane execution initiated at Fri Jul 28 05:18:17 2023

https://github.com/rwest/RMG-Py/actions/runs/5687079457/job/15414935965#step:10:13633
Loading thermodynamics group database from /home/runner/work/RMG-Py/RMG-database/input/thermo/groups...
make: *** [Makefile:61: test-all] Killed
Error: The operation was canceled.
shortly after logging 
running methyl+formaldehyde
Arkane execution initiated at Fri Jul 28 02:18:29 2023

Have you seen this on any of your runs @JacksonBurns ?

rwest commented:
> I'm not sure if it's a consequence of running all the tests is one go (unit, database, and functional) or if it's pytest, or if there's some other way to avoid it, but several of my CI runs (on push) have cancelled in a way that's symptomatic of a test runner machine running out of memory.

Just noticed those jobs were still running nose!

Although it was using the new CI, which runs `make test-all` in one step, the Checkout RMG-Py step of the action is checking out the official main branch, not the branch that is being pushed and tested!?

  /usr/bin/git checkout --progress --force -B main refs/remotes/origin/main
  Switched to a new branch 'main'
  branch 'main' set up to track 'origin/main'.


i.e. when I push to rwest/pytest3 it runs a CI test suite that labels that push as a pass/fail, but it's actually using the code in ReactionMechanismGenerator/main not rwest/pytest3.
 
Perhaps this is a consequence of you trying to make a reusable CI workflow for the purpose of running database tests?

JacksonBurns commented:
Ah yes, I did have a couple seemingly spurious failures but that is definitely the culprit. This is why the `on: pull_request` was working but the forked runs on push to OtherPerson/main were not. Pushing a fix for the CI now.

JacksonBurns commented:
Fixed and pushed to some of the most recent affected branches.

ChrisBNEU commented on Makefile:
```diff
@@ -40,34 +40,23 @@ clean-solver:
 
 test test-unittests:
-	nosetests --nocapture --nologcapture --all-modules -A 'not functional' --verbose --with-coverage --cover-inclusive --cover-erase --cover-html --cover-html-dir=testing/coverage --exe rmgpy arkane
+	python-jl -m pytest -m "not functional not database"
```
I think it needs to be `not functional and not database`. I get an error using pytest 7.1.2 but maybe it's different in other versions.

JacksonBurns replied:
Yes, pushing this now

rwest commented on test/arkane/arkaneKineticsTest.py:
```diff
@@ -52,48 +49,44 @@ def test_give_tlist_for_kineticsjob(self):
-                                   'Obtained values of {0} and {1}'.format(inverse_tlist[1] - inverse_tlist[0],
-                                                                           inverse_tlist[-1] - inverse_tlist[-2]))
+        assert (
+            round(abs(inverse_tlist[1] - inverse_tlist[0] - inverse_tlist[-1] - inverse_tlist[-2]), 7) == 0
```
hang on, shouldn't _A-B≈C-D_  become _(A-B)-(C-D)≈0_ but this is _(A-B-C-D)≈0_  ?  And how many more of these do we need to check? Is there a systematic way to check them all?

JacksonBurns replied:
Ah, thanks for catching that. The original conversion did this in a few places and I fixed _almost_ all of them, apparently. Latest commit (46a8b8b) should have caught the last of them.

I find them using the global search in vscode with regular expressions, finding `round(abs(` where the parentheses then contain more than one `-` symbol.

JacksonBurns commented:
Moved to #2516

rwest commented on .coveragerc:
I missed why .coveragrc is being deleted?

JacksonBurns replied:
Ported all the args to the pytest.ini - did not work with python-jl
